contracts/vesting/src/vesting.rs

```diff
@@ -0,0 +1,559 @@
+        Ok(())
+    }
+
+    fn track_delegation(
+        &self,
+        block_time: Timestamp,
+        mix_identity: IdentityKey,
+        delegation_amount: Coin,
+        storage: &mut dyn Storage,
+    ) -> Result<(), ContractError> {
+        let mut delegations =
+            if let Some(delegations) = get_account_delegations(storage, &self.address) {
+                delegations
+            } else {
+                Vec::new()
+            };
+        delegations.push(DelegationData {
+            mix_identity,
+            amount: delegation_amount.amount,
+            block_time,
+        });
+        set_account_delegations(storage, &self.address, delegations)?;
+        Ok(())
+    }
+
+    fn track_undelegation(
+        &self,
+        mix_identity: IdentityKey,
+        storage: &mut dyn Storage,
+    ) -> Result<(), ContractError> {
+        // This has to exist in storage at this point.
+        let delegations = get_account_delegations(storage, &self.address)
+            .unwrap()
+            .into_iter()
+            .filter(|d| d.mix_identity != mix_identity)
+            .collect();
+        // Since we're always removing the entire delegation we can just drop the key
+
+        Ok(set_account_delegations(
+            storage,
+            &self.address,
+            delegations,
+        )?)
+    }
+}
+
+impl VestingAccount for PeriodicVestingAccount {
+    fn locked_coins(
+        &self,
+        block_time: Option<Timestamp>,
+        env: &Env,
+        storage: &dyn Storage,
+    ) -> Coin {
+        // Returns 0 in case of underflow.
+        Coin {
+            amount: Uint128(
+                self.get_vesting_coins(block_time, env)
+                    .amount
+                    .u128()
+                    .saturating_sub(
+                        self.get_delegated_vesting(block_time, env, storage)
+                            .amount
+                            .u128(),
+                    ),
+            ),
+            denom: DENOM.to_string(),
+        }
+    }
+
+    fn spendable_coins(
+        &self,
+        block_time: Option<Timestamp>,
+        env: &Env,
+        storage: &dyn Storage,
+    ) -> Coin {
+        Coin {
+            amount: Uint128(
+                self.get_balance(storage)
+                    .u128()
+                    .saturating_sub(self.locked_coins(block_time, env, storage).amount.u128()),
+            ),
+            denom: DENOM.to_string(),
+        }
+    }
+
+    fn get_vested_coins(&self, block_time: Option<Timestamp>, env: &Env) -> Coin {
+        let block_time = block_time.unwrap_or(env.block.time);
+        let period = self.get_current_vesting_period(block_time);
+
+        match period {
+            // We're in the first period, or the vesting has not started yet.
+            0 => Coin {
+                amount: Uint128(0),
+                denom: DENOM.to_string(),
+            },
+            1..=7 => Coin {
```

Comment:
Why are period 1-7 treated differently? Do we have only 8 of them in total or something?
If that's the case, perhaps this value should be defined as a constant / state value?

Reply:
Yeah, we always have 8 vesting period, there is a `NUM_VESTING_PERIOD` constant, I'll substitute it here.

Reply:
Turns out I can't, adding a comment instead